### PR TITLE
Allow requesting specific user or group permissions

### PR DESCRIPTION
**Pull request by cmdcolin**
Attempts to address #643 

Optionally accepts a userId or groupId for loadUsers and loadGroups, or no arguments for all users/groups

Also adds exception handling and RestApiDocs

### CODE REVIEW

File: grails-app/controllers/org/bbop/apollo/UserController.groovy [hunk 1]
```diff
@@ -24,98 +24,120 @@ class UserController {
...
-        for (UserOrganismPermission userOrganismPermission in userOrganismPermissionList) {
-            List<UserOrganismPermission> userOrganismPermissionListTemp = userOrganismPermissionMap.get(userOrganismPermission.user.username)
-            if (userOrganismPermissionListTemp == null) {
-                userOrganismPermissionListTemp = new ArrayList<>()
+    @RestApiMethod(description="Load all users",path="/user/loadUsers",verb = RestApiVerb.POST)
+    @RestApiParams(params=[
+            @RestApiParam(name="username", type="email", paramType = RestApiParamType.QUERY)
+            ,@RestApiParam(name="password", type="password", paramType = RestApiParamType.QUERY)
+            ,@RestApiParam(name="userId", type="string", paramType = RestApiParamType.QUERY,description="Optionally only user a specific userId")
+    ])
```
Comment (nathandunn) [inline grails-app/controllers/org/bbop/apollo/UserController.groovy]: Can't believe we didn't have that.  Thanks for adding it.

File: grails-app/controllers/org/bbop/apollo/UserController.groovy [hunk 2]
```diff
@@ -24,98 +24,120 @@ class UserController {
...
-        List<UserOrganismPermission> userOrganismPermissionList = UserOrganismPermission.findAllByOrganismInList(allowableOrganisms as List)
-        for (UserOrganismPermission userOrganismPermission in userOrganismPermissionList) {
-            List<UserOrganismPermission> userOrganismPermissionListTemp = userOrganismPermissionMap.get(userOrganismPermission.user.username)
-            if (userOrganismPermissionListTemp == null) {
-                userOrganismPermissionListTemp = new ArrayList<>()
+    @RestApiMethod(description="Load all users",path="/user/loadUsers",verb = RestApiVerb.POST)
+    @RestApiParams(params=[
+            @RestApiParam(name="username", type="email", paramType = RestApiParamType.QUERY)
+            ,@RestApiParam(name="password", type="password", paramType = RestApiParamType.QUERY)
+            ,@RestApiParam(name="userId", type="string", paramType = RestApiParamType.QUERY,description="Optionally only user a specific userId")
```
Comment (nathandunn) [inline grails-app/controllers/org/bbop/apollo/UserController.groovy]: Is this a string?  Its a long type isn't it?

File: grails-app/controllers/org/bbop/apollo/GroupController.groovy [hunk 3]
```diff
@@ -50,89 +50,101 @@ class GroupController {
...
+                returnArray.put(groupObject)
             }
 
-
-            groupObject.organismPermissions = organismPermissionsArray
-            returnArray.put(groupObject)
+            render returnArray as JSON
+        }
+        catch(Exception e) {
+            response.status=500
```
Comment (nathandunn) [inline grails-app/controllers/org/bbop/apollo/GroupController.groovy]: see comment below about the 500 status vs the enum

File: grails-app/controllers/org/bbop/apollo/UserController.groovy [hunk 4]
```diff
@@ -24,98 +24,120 @@ class UserController {
...
 
-            userObject.organismPermissions = organismPermissionsArray
+                returnArray.put(userObject)
+            }
 
-            returnArray.put(userObject)
+            render returnArray as JSON
+        }
+        catch(Exception e) {
+            response.status=500
```
Comment (nathandunn) [inline grails-app/controllers/org/bbop/apollo/UserController.groovy]: My only comment is that this should be:

        response.status = HttpStatus.SC_INTERNAL_SERVER_ERROR

File: grails-app/controllers/org/bbop/apollo/GroupController.groovy [hunk 5]
```diff
@@ -50,89 +50,101 @@ class GroupController {
     @RestApiParams(params=[
             @RestApiParam(name="username", type="email", paramType = RestApiParamType.QUERY)
             ,@RestApiParam(name="password", type="password", paramType = RestApiParamType.QUERY)
+            ,@RestApiParam(name="groupId", type="string", paramType = RestApiParamType.QUERY,description="Optional only load a specific groupId")
```
Comment (nathandunn) [inline grails-app/controllers/org/bbop/apollo/GroupController.groovy]: should this be type long?

**Comment by nathandunn** [comment]
@cmdcolin  looked good.  . . many improvements in addition to the overall bug. 

 Do you need me to test it or did you already test it?  Otherwise you can merge it as is.  Most comments are relatively minor issues.

**Comment by cmdcolin** [comment]
Should be tested. I tested out the command line api access with curl and used the user panel/group panel without issues

**Pull request merged**